include/swift/Driver/ExperimentalDependencyDriverGraph.h

```diff
@@ -0,0 +1,338 @@
+
+template <typename Key1, typename Key2, typename Value>
+using BiIndexedTwoStageMap =
+    swift::experimental_dependencies::BiIndexedTwoStageMap<Key1, Key2, Value>;
+
+using DeclAspect = swift::experimental_dependencies::DeclAspect;
+
+//==============================================================================
+// MARK: DriverNode
+//==============================================================================
+
+/// A node in the DriverDependencyGraph
+/// Keep separate type from Node for type-checking.
+class DriverNode : public Node {
+
+public:
+  DriverNode(const DependencyKey &key, Optional<std::string> fingerprint)
+      : Node(key, fingerprint) {}
+
+  /// Integrate \p integrand's fingerprint into \p dn.
+  /// \returns true if there was a change requiring recompilation.
+  bool integrateFingerprintFrom(const FrontendNode *integrand) {
+    if (getFingerprint() == integrand->getFingerprint())
+      return false;
+    setFingerprint(integrand->getFingerprint());
+    return true;
+  }
+};
+
+/// A placeholder allowing the experimental system to fit into the driver
+/// without changing as much code.
+class DependencyGraphImpl {
+public:
+  /// Use the status quo LoadResult for now.
+  using LoadResult = typename swift::DependencyGraphImpl::LoadResult;
+};
+
+//==============================================================================
+// MARK: DriverGraph
+//==============================================================================
+
+/// See \ref Node in ExperimentalDependencies.h
+class DriverGraph {
+
+  /// Find nodes, first by the swiftDeps file, then by key.
+  /// Supports searching specific files for a node matching a key.
+  /// Such a search is useful when integrating nodes from a given source file to
+  /// see which nodes were there before integration and so might have
+  /// disappeared.
+  ///
+  /// Some nodes are in no file, for instance a dependency on a Decl in a source
+  /// file whose swiftdeps has not been read yet. For these, the filename is the
+  /// empty string.
+  ///
+  /// Don't add to this collection directly; use \ref addToMap
+  /// instead because it enforces the correspondence with the swiftFileDeps
+  /// field of the node.
+  /// TODO: Fix above comment
+  ///
+  /// Sadly, cannot use an optional string for a key.
+  using NodeMap =
+      BiIndexedTwoStageMap<std::string, DependencyKey, DriverNode *>;
+  NodeMap nodeMap;
+
+  /// Since dependency keys use baseNames, they are coarser than individual
+  /// decls. So two decls might map to the same key. Index the second level by
+  /// file. Very useful to find a Node given the key, for instance when finding
+  /// all the nodes affected by a change in one.
+
+  /// Tracks def-use relationships by DependencyKey.
+  std::unordered_map<DependencyKey, std::unordered_set<DependencyKey>>
+      usesByDef;
+
+  // Supports requests from the driver to getExternalDependencies.
+  std::unordered_set<std::string> externalDependencies;
```

Comment:
Long-term nitpick: apart from using the `std::` data structures vs. the LLVM data structures, it's probably worth preserving order to guarantee determinism anyway.

Reply:
I assume the comment applies to all uses of the std::unordered_* containers. Good thought. @jckarter pointed out that the LLVM data structures don't provide stable pointers, so I think there would be additional lookups to use them. That's a good direction to investigate when performance tuning happens.

W.r.t. to determinism, that's a good point. Let's see.... The unordered containers should still be deterministic for the same inputs, so you must be thinking that the inputs can vary, is that right? And then, the compilation order might change, which could excite any lurking compilation-order-dependencies. Is that the bigger picture?

Reply:
I'm not sure there's even a guarantee that the unordered containers are deterministic for the same input—the Swift ones aren't, for example—but at the very least it won't be the *same* order across C++ stdlibs, which makes it hard to write platform-independent tests. Either a sorted container (`std::set`) or an ordered set (`llvm::SetVector`) would avoid this particular problem.

(We already know we can have different internal behavior when building things in a different order or in parallel.)